### PR TITLE
feat: add ButtonBase react component

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This PR introduces a new ButtonBase component to the design system. The ButtonBase serves as the foundational component for all button variants, providing essential functionality and consistent styling.

Key features:
- Three size variants (sm: 32px, md: 40px, lg: 48px)
- Full width support
- Loading state with optional loading text
- Icon support (start/end positions)
- Custom accessories support
- Disabled state
- AsChild prop for polymorphic rendering
- Comprehensive test coverage
- Storybook documentation with examples

## **Related issues**

Fixes: #140

## **Manual testing steps**

1. Run Storybook (`yarn storybook`)
2. Navigate to the ButtonBase component
3. Test the following features:
   - Different sizes (sm, md, lg)
   - Loading state with and without loading text
   - Icon placement (start/end)
   - Custom accessories
   - Disabled state
   - Full width mode
   - AsChild functionality with anchor tags
4. Verify styles and transitions
5. Test keyboard navigation and accessibility

### **Before**
N/A - New component

### **After**

https://github.com/user-attachments/assets/853725c2-65f1-4522-87ef-4b3a9f1993f8

Adding animation

https://github.com/user-attachments/assets/80550c38-a3f1-46c5-aa29-026b09374393

100% test coverage

Screenshot 2024-12-10 at 5 23 47 PM

## **Pre-merge author checklist**

- [x] I've followed MetaMask Contributor Docs
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests (comprehensive test suite with 97% coverage)
- [x] I've documented my code using JSDoc format
- [ ] I've applied the right labels on the PR

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR
- [ ] I confirm that this PR addresses all acceptance criteria

### CODE REVIEW

File: packages/design-system-react/src/components/button-base/ButtonBase.constants.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Remove this file if it's not needed
+import { ButtonBaseSize } from './ButtonBase.types';
+
+export const BUTTONBASE_CLASSMAP = {};
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.constants.ts:4]: Remove this
```suggestion
```

File: packages/design-system-react/src/components/button-base/ButtonBase.constants.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Remove this file if it's not needed
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.constants.ts:1]:
```suggestion
```

File: packages/design-system-react/src/components/button-base/ButtonBase.constants.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+import { ButtonBaseSize } from './ButtonBase.types';
+
+export const BUTTONBASE_CLASSMAP = {};
+
+export const SIZE_CLASS_MAP: Record<ButtonBaseSize, string> = {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.constants.ts:6]:
```suggestion
export const BUTTON_BASE_SIZE_CLASS_MAP: Record<ButtonBaseSize, string> = {
```

File: packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    docs: {
+      page: README,
+    },
+  },
+  argTypes: {
+    size: {
+      control: 'select',
+      options: Object.values(ButtonBaseSize),
+    },
+  },
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx:21]: Add all the different argTypes here

File: packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  render: () => (
+    <div className="flex gap-2">
+      <ButtonBase size={ButtonBaseSize.Sm}>Small</ButtonBase>
+      <ButtonBase size={ButtonBaseSize.Md}>Medium</ButtonBase>
+      <ButtonBase size={ButtonBaseSize.Lg}>Large</ButtonBase>
+    </div>
+  ),
+};
+
+export const FullWidth: Story = {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx:43]: Make sure the stories and titles in readme are a direct match to the prop name just in PascalCase

File: packages/design-system-react/src/components/button-base/ButtonBase.test.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    expect(accessory).toHaveClass('ml-2');
+  });
+
+  it('applies disabled state', () => {
+    render(<ButtonBase isDisabled>Disabled Button</ButtonBase>);
+    const button = screen.getByRole('button');
+    expect(button).toBeDisabled();
+    expect(button).toHaveClass('opacity-50', 'cursor-not-allowed');
+  });
+});
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.test.tsx:97]: Ensure 100% coverage of props

File: packages/design-system-react/src/components/button-base/ButtonBase.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      SIZE_CLASS_MAP[size],
+      // Full width
+      isFullWidth && 'w-full',
+      // Disabled state
+      isDisabled && 'opacity-50 cursor-not-allowed',
+      // Loading state
+      isLoading && 'cursor-wait',
+      // Custom classes
+      className,
+    );
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.tsx:50]: I've created an issue to improve the way we apply tailwind classnames to components and possibly adopting the [class variance authority](https://cva.style/docs) approach which may align better with how engineers that like CSS in JS use Tailwind

File: packages/design-system-react/src/components/button-base/ButtonBase.types.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import type { ComponentProps } from 'react';
+
+import type { IconName, IconProps } from '../icon';
+
+export enum ButtonBaseSize {
+  Sm = 'sm',
+  Md = 'md',
+  Lg = 'lg',
+}
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.types.ts:9]: Add JsDoc comments here with pixel values

File: packages/design-system-react/src/components/button-base/ButtonBase.types.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+export type ButtonBaseProps = ComponentProps<'button'> & {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.types.ts:11]: Make sure all types that are optional use the "optional prop" language present in other components

File: packages/design-system-react/src/components/button-base/README.mdx [hunk 10]
```diff
@@ -0,0 +1,73 @@
+import { Controls, Canvas, Meta } from '@storybook/blocks';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/README.mdx:1]: Make sure the docs follow the prop name format and use the enums

File: packages/design-system-react/src/components/button-base/index.ts [hunk 11]
```diff
@@ -0,0 +1,2 @@
+export { ButtonBase } from './ButtonBase';
+export type { ButtonBaseProps } from './ButtonBase.types';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/index.ts:2]: Make sure we export all types from index

File: packages/design-system-react/src/components/index.ts [hunk 12]
```diff
@@ -13,3 +13,6 @@ export type { TextProps } from './text';
 export { Icon } from './icon';
 export { IconName, IconSize, IconColor } from './icon';
 export type { IconProps } from './icon';
+
+export { ButtonBase } from './button-base';
+export type { ButtonBaseProps } from './button-base';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/index.ts:18]: Make sure we export all types from index

File: packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    endAccessory: {
+      control: 'text',
+      description:
+        'Optional prop for a custom element to show at the end of the button',
+    },
+    isDisabled: {
+      control: 'boolean',
+      description: 'Optional prop that when true, disables the button',
+    },
+  },
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/ButtonBase.stories.tsx:91]: Adds controls to story and improves component API section

Screenshot 2024-12-10 at 4 31 24 PM
Screenshot 2024-12-10 at 4 31 40 PM

File: packages/design-system-react/src/components/button-base/ButtonBase.test.tsx (hunk 6 above)
Comment (georgewrmarshall) [reply packages/design-system-react/src/components/button-base/ButtonBase.test.tsx:122]: Screenshot 2024-12-10 at 4 32 41 PM

File: packages/design-system-react/src/components/button-base/README.mdx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+```tsx
+import { ButtonBase } from '@metamask/design-system-react';
+
+<ButtonBase onClick={() => {}}>Button Base</ButtonBase>;
+```
+
+### Size
+
+ButtonBase supports three sizes: `sm`, `md`, and `lg`.
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/button-base/README.mdx:21]: Add `ButtonBaseSize` enum

File: packages/design-system-react/src/components/button-base/ButtonBase.types.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
+import type { ComponentProps } from 'react';
+
+import type { MakePropsOptional } from '../../types/make-props-optional';
+import type { IconName, IconProps } from '../icon';
+
+export enum ButtonBaseSize {
```
Comment (brianacnguyen) [inline packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: Shouldnt this be `ButtonSize`?

Comment (brianacnguyen) [inline packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: Shouldnt this be ButtonSize?

**Review by brianacnguyen — CHANGES_REQUESTED** [review]
a small change requested

File: packages/design-system-react/src/components/button-base/ButtonBase.types.ts (hunk 15 above)
Comment (georgewrmarshall) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: The enum name `ButtonBaseSize` follows our established naming pattern of `ComponentName` + `Prop` = `EnumName`. 

In this case:
- Component name: `ButtonBase`
- Prop name: `size`
- Resulting enum name: `ButtonBaseSize`

This naming convention helps maintain consistency across our component library and makes the relationship between components, props, and their associated types more explicit.

Comment (georgewrmarshall) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: Issue here https://github.com/MetaMask/metamask-design-system/issues/292

Comment (brianacnguyen) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: are we intended for the user to import `ButtonSize` or `ButtonBaseSize`? I assume it should be `ButtonSize`. If that's the case, are you intending to map the `ButtonSize` to `ButtonBaseSize`?

Comment (georgewrmarshall) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: Yes see the draft of `ButtonPrimary` we are exporting `ButtonBaseSize` as `ButtonPrimarySize` and importing it for usage with `ButtonPrimary` 

https://github.com/MetaMask/metamask-design-system/pull/289/files#diff-ff7a98d6e5b8ccd27ebdec0b0c9a70b9551ef9651863121cc65b58ab08fefc90R3

https://github.com/MetaMask/metamask-design-system/pull/289/files#diff-fe3e24d7920d4e19ada30de93b6830e5439b0ff9d69da9a1402101ed4ca92a11R4

Comment (brianacnguyen) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: sounds good. I just want to make sure the user doesnt need to do something like 
`<Button size={ButtonBaseSize.Lg}/>`

**Review by brianacnguyen — APPROVED** [review]

File: packages/design-system-react/src/components/button-base/ButtonBase.types.ts (hunk 15 above)
Comment (georgewrmarshall) [reply packages/design-system-react/src/components/button-base/ButtonBase.types.ts:6]: Yup, good call we definitely want to align the enum with the exposed component.

**Review by brianacnguyen — APPROVED** [review]

**Pull request merged**